File: backend/src/BiteRight.Application/Dtos/Products/DetailedProductDto.cs
// # ==============================================================================
// # Solution: BiteRight
// # File: DetailedProductDto.cs
// # Author: Łukasz Sobczak
// # Created: 25-02-2024
// # ==============================================================================

#region

using System;
using BiteRight.Application.Dtos.Categories;
using BiteRight.Application.Dtos.Currencies;
using BiteRight.Application.Dtos.Units;
using BiteRight.Domain.Languages;
using BiteRight.Domain.Products;

#endregion

namespace BiteRight.Application.Dtos.Products;

public class DetailedProductDto
{
    public DetailedProductDto(
        Guid id,
        string name,
        string description,
        double? priceValue,
        CurrencyDto? priceCurrency,
        ExpirationDateKindDto expirationDateKind,
        DateOnly? expirationDateValue,
        CategoryDto categoryDto,
        DateTime addedDateTime,
        double amountCurrentValue,
        double amountMaxValue,
        UnitDto amountUnit,
        bool disposedStateValue,
        DateTime? disposedStateDateTime
    )
    {
        Id = id;
        Name = name;
        Description = description;
        PriceValue = priceValue;
        PriceCurrency = priceCurrency;
        ExpirationDateKind = expirationDateKind;
        ExpirationDateValue = expirationDateValue;
        Category = categoryDto;
        AddedDateTime = addedDateTime;
        AmountCurrentValue = amountCurrentValue;
        AmountMaxValue = amountMaxValue;
        AmountUnit = amountUnit;
        DisposedStateValue = disposedStateValue;
        DisposedStateDateTime = disposedStateDateTime;
    }

    public Guid Id { get; set; }
    public string Name { get; set; }
    public string Description { get; set; }
    public double? PriceValue { get; set; }
    public CurrencyDto? PriceCurrency { get; set; }
    public ExpirationDateKindDto ExpirationDateKind { get; set; }
    public DateOnly? ExpirationDateValue { get; set; }
    public CategoryDto Category { get; set; }
    public DateTime AddedDateTime { get; set; }
    public double AmountCurrentValue { get; set; }
    public double AmountMaxValue { get; set; }
    public UnitDto AmountUnit { get; set; }
    public bool DisposedStateValue { get; set; }
    public DateTime? DisposedStateDateTime { get; set; }

    public static DetailedProductDto FromDomain(
        Product product,
        LanguageId languageId
    )
    {
        return new DetailedProductDto(
            product.Id,
            product.Name,
            product.Description,
            product.Price?.Value,
            product.Price is null ? null : CurrencyDto.FromDomain(product.Price.Currency),
            (ExpirationDateKindDto)product.ExpirationDate.Kind,
            product.ExpirationDate.Value,
            CategoryDto.FromDomain(product.Category, languageId),
            product.AddedDateTime,
            product.Amount.CurrentValue,
            product.Amount.MaxValue,
            UnitDto.FromDomain(product.Amount.Unit, languageId),
            product.DisposedState.Value,
            product.DisposedState.DateTime
        );
    }
}

File: mobile/app/src/main/java/com/sobczal2/biteright/dto/products/DetailedProductDto.kt
package com.sobczal2.biteright.dto.products

import com.google.gson.annotations.SerializedName
import com.sobczal2.biteright.dto.categories.CategoryDto
import com.sobczal2.biteright.dto.currencies.CurrencyDto
import com.sobczal2.biteright.dto.units.UnitDto
import java.time.Instant
import java.time.LocalDate
import java.time.LocalDateTime
import java.util.UUID

data class DetailedProductDto(
    @SerializedName("id") val id: UUID,
    @SerializedName("name") val name: String,
    @SerializedName("description") val description: String,
    @SerializedName("priceValue") val priceValue: Double?,
    @SerializedName("priceCurrency") val priceCurrency: CurrencyDto?,
    @SerializedName("expirationDateKind") val expirationDateKind: ExpirationDateKindDto,
    @SerializedName("expirationDateValue") val expirationDateValue: LocalDate?,
    @SerializedName("category") val category: CategoryDto,
    @SerializedName("addedDateTime") val addedDateTime: Instant,
    @SerializedName("amountCurrentValue") val amountCurrentValue: Double,
    @SerializedName("amountMaxValue") val amountMaxValue: Double,
    @SerializedName("amountUnit") val amountUnit: UnitDto,
    @SerializedName("disposedStateValue") val disposedStateValue: Boolean,
    @SerializedName("disposedStateDateTime") val disposedStateDateTime: Instant?
)